File: LeetCode/Medium/1061-lexicographically-smallest-equivalent-string/1061-lexicographically-smallest-equivalent-string.py
class Solution:
    def smallestEquivalentString(self, s1: str, s2: str, baseStr: str) -> str:
        targets: dict[str, str] = {x : x for x in string.ascii_lowercase}
        # print(", ".join(f"{k}->{v}" for k,v in targets.items()))

        def find(c: str) -> str:
            if c != targets[c]:
                targets[c] = find(targets[c])
            return targets[c]

        def union(x: str, y: str):
            target_x = find(x)
            target_y = find(y)

            # print(f"{target_x} {target_y}")

            if target_x < target_y:
                targets[target_y] = target_x
            else:
                targets[target_x] = target_y

        for x, y in zip(s1, s2):
            union(x, y)
        # print(", ".join(f"{k}->{v}" for k,v in targets.items()))


        base = list(baseStr)
        for i in range(len(base)):
            base[i] = find(base[i])
        return "".join(base)

File: LeetCode/Medium/1061-lexicographically-smallest-equivalent-string/1061-lexicographically-smallest-equivalent-string.cpp
class Solution {
    std::array<char, 26> parent;

    int getIndex(char c) const {
        return c - 'a';
    }

public:
    Solution()
    {
        for (auto i = 0; i < 26; ++i) {
            parent[i] = 'a' + i;
        }
    }

    char find(char c)
    {
        auto idx = getIndex(c);
        if (parent[idx] == c)
        {
            return c;
        }

        return parent[idx] = find(parent[idx]);
    }

    void unite(char c1, char c2)
    {
        auto root1 = find(c1);
        auto root2 = find(c2);

        if (root1 < root2)
        {
            parent[getIndex(root2)] = root1;
        }
        else
        {
            parent[getIndex(root1)] = root2;
        }
    }

    string smallestEquivalentString(string s1, string s2, string baseStr) {
        for (auto [c1, c2] : std::views::zip(s1, s2)) {
            unite(c1, c2);
        }

        return baseStr
                | std::views::transform([this](char c) { return find(c); })
                | std::ranges::to<std::string>();
    }
};